### ISSUE TITLE
inconsistence when reading midi in score note and fretboard show

### ISSUE THREAD

**Issue by sugizo** [helge17/tuxguitar#150]
![image](https://github.com/helge17/tuxguitar/assets/2418295/622a600a-93b1-4d9e-b01f-afe4da379229)

**background**
the midi is about to learn C Ionian Scale or C Major Scale
that consist the note :
C4 D4 E4 F4 G4 A4 B4 C5 B4 A4 G4 F4 E4 D4 C4

**problem** 
on the note score and fretboard show from image above
the note score and fretobard show after C5 which are 
B4 G4 E4 and C4
consist more than 1 notes, that is incorrect, 
it's should only consist one note

**expectation**
the correct note score should be like on musescore 
that show only 1 note 
when load the same C_Ionian_scale.mid file
![image](https://github.com/helge17/tuxguitar/assets/2418295/bea25ad2-e579-49ef-ab7a-7b79e3eb9a2f)

**n.b.**
could send the C_Ionian_scale.mid file, 
if you want to test your self

thanks and best regards

**Comment by guiv42** [comment]
yes I would appreciate if you can provide the .mid file, or else I have no means to reproduce & debug (you'll probably need to zip it to include it in a comment here)

**Comment by guiv42** [comment]
Thank you. The good news is I can reproduce the issue with this file.
I'm currently working on another topic (#142), once I have finished I will have a look at this issue

**Comment by sugizo** [comment]
still in same problem report above but, 
this case : the note is missing
![Screenshot 2023-11-10 220202](https://github.com/helge17/tuxguitar/assets/2418295/e413a39d-ac4a-4971-b9e3-07e17ec61601)

as you can see on measure 8 and 9 the note is less than, 
the same midi file open with musescore
![Screenshot 2023-11-10 220227](https://github.com/helge17/tuxguitar/assets/2418295/570c17db-9c92-4d7e-955a-7d9f2ebad401)

detail
in measure 8 
tux guitar show : 7 note
while
musescore show : 10 note

in measure 9  
tux guitar show : 2 note
while
musescore show : 8 note

best regards
[Ayumi Hamasaki - You.zip](https://github.com/helge17/tuxguitar/files/13320735/Ayumi.Hamasaki.-.You.zip)

**Comment by guiv42** [comment]
@sugizo : could you please describe how you produced the C_Ionian_scale file?
I decoded the binary file by hand to understand where the problem could be, and it does not seem consistent with the specification I found [here](https://www.cs.cmu.edu/~music/cmsip/readings/Standard-MIDI-file-format-updated.pdf).
By the way, when I create in TG a C ionian scale as depicted above, then export in midi format, I can decode the binary file manually. So most probably TG was developed in consistency with the spec I have found.
The question behind: is TG using the right specification to encode/decode midi?

**Comment by guiv42** [comment]
a beginning of an explanation for the C Ionian scale. When comparing manually the file provided above with one correct Ionian scale produced by TuxGuitar
- TuxGuitar encodes end of notes with "note OFF" events (that makes sense!)
- the attached file encodes end of notes with events "note ON, velocity zero", effectively muting the note

In the file provided above, when going up the scale, each time the note changes 2 events are included: "mute current note, start new note". When going down the scale the order is inverted: "start new note, mute preceding note". It seems this very last event is not properly understood by TG.

**Comment by sugizo** [comment]
nice catch for root cause analysis in detail (goint up n going down in scale)

have two questions :
1. what tools you use to spot that (in root cause analysis above) ?
2. what should user do, because many users, just download the midi on internet not producing their own midi ?

best regards

**Comment by guiv42** [comment]
1. what tool: hexdump... then just decoding manually the binary .mid file
2. users can't do anything. Now I have understood *when* it happens I will dig in the code to understand *why*. With the objective to fix the issue.

@sugizo : just one question, which tool did you use to generate those midi files? I will need to generate a few of them for testing purposes

**Comment by sugizo** [comment]
the one sent in this github generated by python module : 
pretty_midi

could send another midi if you want (another scale, arpeggio already tested wrong when open in tux guitar (generated by pretty_midi) ), 
or 
could sent the source for generated those midi using python module : pretty midi (jupyter notebook or use google colab)

n.b.
for learning purpose, tested several too all in python : 
miditoolkit, midiutil, mido, music21, musicpy
the flow : 
read json source for midi number then generate using those python module

![image](https://github.com/helge17/tuxguitar/assets/2418295/cd9e2858-794a-48f7-8897-4bed361f234a)

**Comment by sugizo** [comment]
![image](https://github.com/helge17/tuxguitar/assets/2418295/a3a15276-4ac3-462c-a75a-78cc058b25af)

could anybody help me translate the meaning of this hexdump result ?
the important part :
- note on
- note off
- velocity 0
- velocity 100
- pitch e.g. C4
- instrument, e.g. accoustic grand piano
- duration

already test with another instrument e.g. violin, cello, 
seems got the same result

**Comment by guiv42** [comment]
it seems one needs to register to have access to the official midi specifications.
I found [this doc](https://www.cs.cmu.edu/~music/cmsip/readings/Standard-MIDI-file-format-updated.pdf) is a good starting point. It's the one I used to decode the Ionian scale example above. Take this as guidance only, I needed to extrapolate/guess a bit the detailed coding. Ionian scale example is not 100% compliant with this doc (even going up the scale, where TuxGuitar interprets it correctly)

**Comment by sugizo** [comment]
thanks for the ref, trying to learn the midi doc first

C_Ionian_scale.mid analyzed with 3 different python library mentions above

pretty_midi
![Screenshot 2023-11-16 030244](https://github.com/helge17/tuxguitar/assets/2418295/d09297e7-06e1-420b-859f-5b8fe0ab4895)

mido
![Screenshot 2023-11-16 030355](https://github.com/helge17/tuxguitar/assets/2418295/66876ed3-ce35-469a-937b-ba5dc2a0a9c4)

miditoolkit
![Screenshot 2023-11-16 030456](https://github.com/helge17/tuxguitar/assets/2418295/0af43715-9662-4e76-94d1-f56054607ee3)

mido result, same like your deduction mention above, 
```
the attached file encodes end of notes with events "note ON, velocity zero", effectively muting the note

In the file provided above, 
when going up the scale, each time the note changes 2 events are included: "mute current note, start new note". 
When going down the scale the order is inverted: "start new note, mute preceding note".
```

but the same file, can be read correctly in musescore

**Comment by sugizo** [comment]
midi generator :
using note_on, note_off :
- midiutil
- mido
- music21
- musicpy

using note_on with velocity 0 for note_off, and incorrect order for note_off when note going down :
- pretty_midi
- miditoolkit

all tools above, analyzed with mido

is tux guitar able to generate midi for scale, chord, etc, 
without user know about how to write musical score ?

just like, tools tested above, 
all midi is created with midi number (e.g. 64) 
or 
note with octave (e.g. C4)

**Comment by guiv42** [comment]
Ionian scale example:
> but the same file, can be read correctly in musescore

Yes it looks TuxGuitar import gets confused by this specific sequence. As mentioned above I'll have a look when I'm done with the tunings dialog.



> is tux guitar able to generate midi for scale, chord, etc,
> without user know about how to write musical score ?

TuxGuitar provides a chord editor (default shortcut 'a'), I don't know about scales (don't know all features of this tool yet). Anyway, TuxGuitar has been designed as a **tablature editor**, not a general music educational software, not a generic midi software either, so I'm not surprised it does not provide such features. Other more generic tools can do this much better, and I'm personally perfectly OK with that (keeping focused on guitar).

**Comment by guiv42** [comment]
There are 2 very different problems there

1. C_Ionian_scale.zip
My analysis above is confirmed: the order of simultaneous events within midi file had an impact on the import function.
Should be fixed by pull request #155

2. Ayumi Hamasaki - You.zip
Refer to screenshots in [this post above](https://github.com/helge17/tuxguitar/issues/150#issuecomment-1805916662)
When I import this midi file with TuxGuitar 1.6.0 I do not get the same result.
Here is a screenshot of measures 8 and 9 after import (with parameter "transpose" set to zero):
![Screenshot_20231121_220819](https://github.com/helge17/tuxguitar/assets/129443524/d1587fe5-617c-4436-a45b-4f23c7929dfd)

In this case, number of notes seems globally consistent with what musescore displays (considering tied notes).
However, I do not understand at all the tablature part. For an unknown reason, track tuning is defined to: A# A# A# A# A# A# (?!?)
And when I try to change to standard guitar tuning, then some notes disappear (?!?)
@sugizo : is this consistent with your observations?

**Comment by sugizo** [comment]
number one, is just an info, 
so 
i'll reply the number two problem :
yes, exactly is same with my observation report above 

the tab finger position different, 
perhaps because we have different tunning, 

mine using major third custom : 
Ab2 c3 e3 ab3 c4 e4

don't know much about the music score notation ("tied notes", you mention above)
but 
the output sound is missing some note in tux guitar (this is my point of view problem)

not sure about the tuning define to 
A# A# A# A# A# A# 
by default, 
as i observed before the default instrument for that midi is piano

**Comment by guiv42** [comment]
OK I got it.

First thing: tied notes matter and cannot be ignored. A midi file encodes *sound*, and there can be several different and equivalent representations of this *sound* on a *score*. So, one cannot expect 2 tools to display the same *score* while importing the same midi file (not even to display the same number of notes).

That being said, in this case it's true some notes are lost. But it's **not** a TuxGuitar issue.
The "Ayumi Hamasaki - You" midi file is not written for guitar, and is way too high. When you import the midi file it's OK, because TuxGuitar chooses an arbitrary tuning so that all notes can be reached. This leads to this somehow stupid tuning.
But when you change tuning, some notes are lost... just because they become too high for the fretboard and cannot be played.
Keep in mind that guitar plays 1 octave below the standard treble clef (this is why notes do not appear at the same height in musescore and TuxGuitar).

I don't know where this file comes from, but you need to transpose it somehow to display it correctly in TuxGuitar. Either you can do this outside of TuxGuitar, or you can do it this way:
- File/import/import midi, select your file
- in the popup, transpose -12. This displays some inconsistent fret numbers on the tablature
- change tuning

**Comment by guiv42** [comment]
With pull request #156, importing "Ayumi Hamasaki - You" midi file with transposition of -12 works fine: resulting score correctly fits in standard tuning.

**Comment by sugizo** [comment]
why tested on snapshot 2023-11-19, not have the same result as your observation ?

**in measure 8**
tux guitar show from your observation : 8 notes

**in measure 9**
tux guitar show from your observation : 4 notes

**test result file > import > import midi**
![image](https://github.com/helge17/tuxguitar/assets/2418295/1dd18a41-850b-4b61-a6d6-bfe519581173)

**but can open it when using file > history**
![image](https://github.com/helge17/tuxguitar/assets/2418295/089b56e4-3914-485d-ad77-b15ec027c2a6)

**comment**
when using file history result is work as expected '2023-11-19 snapshot'
while the note not same like musescore but
the note count and output sound is same

**steps** 
```
powershell -Command "Remove-Item C:\tuxguitar-*\ -Recurse"
cd %homepath%\Downloads
curl -LC - https://github.com/helge17/tuxguitar/releases/download/2023-11-19-snapshot/tuxguitar-2023-11-19-snapshot-windows-swt-x86_64.zip -o tuxguitar.zip
powershell -Command "Expand-Archive -LiteralPath %HOMEPATH%\Downloads\tuxguitar.zip -DestinationPath C:\tuxguitar"
powershell -Command "Copy-Item -Path 'C:\tuxguitar\*' -Destination 'C:\' -Recurse"
powershell -Command "Remove-Item C:\tuxguitar -Recurse"
powershell -Command "Remove-Item %HOMEPATH%\Downloads\tuxguitar.zip"
C:\tuxguitar-2023-11-19-snapshot-windows-swt-x86_64\tuxguitar.exe
```

**Comment by guiv42** [comment]
Please be patient. Refer to [commit history](https://github.com/helge17/tuxguitar/commits/master): the 2 fixes for midi import were implemented on Nov. 21st and 22nd, so *after* 2023-11-19 was published.

**Comment by sugizo** [comment]
>  Keep in mind that guitar plays 1 octave below the standard treble clef (this is why notes do not appear at the same height in musescore and TuxGuitar).

is this the reason why : 
in tux guitar tuning standard is E3 A3 D4 G4 B4 E5
while 
the wiki told the 6 strings standard tuning is E2 A2 D3 G3 B3 E4
notice the octave different
?

**ref**
https://en.wikipedia.org/wiki/Guitar_tunings

**another question**
File/import/import midi, have the popup, to transpose 

why this feature not available on :
File/open/ 
and 
File/history 
that 
select midi file
?

**steps**
```
powershell -Command "Remove-Item C:\tuxguitar-*\ -Recurse"
cd %homepath%\Downloads
curl -LC - https://github.com/helge17/tuxguitar/releases/download/2023-11-23-snapshot/tuxguitar-2023-11-23-snapshot-windows-swt-x86_64.zip -o tuxguitar.zip
powershell -Command "Expand-Archive -LiteralPath %HOMEPATH%\Downloads\tuxguitar.zip -DestinationPath C:\tuxguitar"
powershell -Command "Copy-Item -Path 'C:\tuxguitar\*' -Destination 'C:\' -Recurse"
powershell -Command "Remove-Item C:\tuxguitar -Recurse"
powershell -Command "Remove-Item %HOMEPATH%\Downloads\tuxguitar.zip"
C:\tuxguitar-2023-11-23-snapshot-windows-swt-x86_64\tuxguitar.exe
```

**result**
- velocity 0 replace note off (c_ionian.mid) : fixed
- missing note on midi file tux guitar compare to musescore (ayumi hamasaki - you.mid) : fixed

best regards

**Comment by guiv42** [comment]
> is this the reason why :
in tux guitar tuning standard is E3 A3 D4 G4 B4 E5
while
the wiki told the 6 strings standard tuning is E2 A2 D3 G3 B3 E4

No, that is another long-lived bug we could not fix up to now, see #37 

> File/import/import midi, have the popup, to transpose
why this feature not available on :
File/open/

I think it's been like this forever, and I find this consistent: "import" menu item proposes you more options to transform input data than "open" does. Or why would these 2 different menu items exist?

So, if everything is fixed, I suggest closing this issue if it's OK for you.
